File: Point.swift
//
//  Point.swift
//  TestEngine
//
//  Created by Vladas Zakrevskis on 7/3/17.
//  Copyright © 2017 VladasZ. All rights reserved.
//

import Foundation

infix operator •: AdditionPrecedence

func •(left: Point, right: Point) -> Double {
    
    return left.x * right.x + left.y * right.y
}

struct Point {
    
    var x: Double = 0
    var y: Double = 0
    
    var angle: Double { return .pi - atan2(x, y) }
    var isZero: Bool { return x == 0 && y == 0 }
    var length: Double { return sqrt(pow(abs(x), 2) + pow(abs(y), 2)) }

     static var one: Point { return Point(0, 1) }
    
    var normalized: Point {
        
        if self.isZero { return Point.one }
        return self * (1 / length)
    }
    
    init() { }
    
    init(_ x: Double, _ y: Double) {
    
        self.x = x
        self.y = y
    }
    
    static func -(left: Point, right: Point) -> Point {
        
        return Point(left.x - right.x, left.y - right.y)
    }
    
    static func +(left: Point, right: Point) -> Point {
    
        return Point(left.x + right.x, left.y + right.y)
    }
    
    static func += ( left: inout Point, right: Point) {
        
        left.x += right.x
        left.y += right.y
    }
    
    static func *(left: Point, right: Double) -> Point {
        
        return Point(left.x * right, left.y * right)
    }
    
    func distanceTo(_ point: Point) -> Double {
        
        let xDist = x - point.x
        let yDist = y - point.y
        return sqrt((xDist * xDist) + (yDist * yDist))
    }
    
    func withLength(_ length: Double) -> Point {
        
        return normalized * length
    }
    
    func angleWith(_ point: Point) -> Double {
        
        return (self • point) / self.length / point.length
    }
    
    func projectionTo(_ point: Point) -> Double {
        
        return self.length * angleWith(point)
    }
}

extension Point : CustomStringConvertible {
    
    var description: String { return "Point: x - \(x), y - \(y)" }
}

File: source/Engine/Geometry/Primitives/Point.cpp
//
//  Point.cpp
//  TestEngine
//
//  Created by Vladas Zakrevskis on 8/16/17.
//  Copyright © 2017 VladasZ. All rights reserved.
//

#include "Point.hpp"
#include <math.h>

Point::Point(float x, float y) : x(x), y(y) { }

Point::Point(Direction direction, float length) {
    switch (direction) {
        case Direction::Left:  x = -length; y =  0;      break;
        case Direction::Right: x =  length; y =  0;      break;
        case Direction::Up:    x =  0;      y = -length; break;
        case Direction::Down:  x =  0;      y =  length; break;
        default: break;
    }
}

Point Point::onCircle(float radius, float angle, const Point &center) {
    return Point((radius / 2) * cos(angle) + center.x, (radius / 2) * sin(angle) + center.y);
}

float Point::angle() const {
    return atan2(y, x);
}

bool Point::isZero() const {
    return x == 0 && y == 0;
}

float Point::length() const {
    return sqrt(x * x + y * y);
}

Point Point::withLength(float length) const {
    float ratio = length / this->length();
    return Point(x * ratio, y * ratio);
}

Direction Point::directionX() const {
    return x > 0 ? Direction::Right : Direction::Left;
}

Point Point::operator +(const Point &point) const {
    return Point(x + point.x, y + point.y);
}

void Point::operator +=(const Point &point) {
    x += point.x;
    y += point.y;
}

Point Point::operator -(const Point &point) const {
    return Point(x - point.x, y - point.y);
}

void Point::operator -=(const Point &point) {
    x -= point.x;
    y -= point.y;
}

Point Point::operator *(float value) const {
    return Point(x * value, y * value);
}

void Point::operator *=(float value) {
    x *= value;
    y *= value;
}
